### PR TITLE
Improve coverage

### PR DESCRIPTION
**Pull request by briandesilva**
Increased unit test coverage for

- differentiation
- optimization
- utils
- library formation
- SINDy

along with some other minor changes

- Link readthedocs documentation in README
- Removed extraneous comments
- Made `get_feature_names` an abstract method of `BaseFeatureLibrary`
- Corrected comment typos
- Added `complexity` method to `BaseOptimizer`
- Removed unused debug function
- Fix error types raised in `process_multiple_trajectories`

### CODE REVIEW

File: test/differentiation/test_differentiation_methods.py [hunk 1]
```diff
@@ -111,6 +113,28 @@ def test_centered_difference_dim():
...
+    "data",
+    [
+        pytest.lazy_fixture("data_derivative_1d"),
+        pytest.lazy_fixture("data_derivative_2d"),
+    ],
+)
+def test_smoothed_finite_difference(data):
+    x, x_dot = data
+    smoothed_centered_difference = SmoothedFiniteDifference()
+    smoothed_centered_difference(x)
```
[the remark this answers is not in the document]
Comment (briandesilva) [reply test/differentiation/test_differentiation_methods.py]: Fixing in next commit

File: test/feature_library/test_feature_library.py [hunk 2]
```diff
@@ -72,3 +92,71 @@ def test_output_shape(data_lorenz, library, shape):
...
+
+
+# Catch-all for various combinations of options and
+# inputs for polynomial features
+def test_polynomial_options(data_lorenz):
+    x, t = data_lorenz
+
+    # Check sparse inputs
+    library = PolynomialLibrary()
+    library.fit_transform(csc_matrix(x))
```
[the remark this answers is not in the document]
Comment (briandesilva) [reply test/feature_library/test_feature_library.py]: Fixed in upcoming commit

File: test/test_pysindy.py [hunk 3]
```diff
@@ -323,3 +326,97 @@ def test_score_discrete_time_multiple_trajectories(
...
+        pytest.lazy_fixture("data_lorenz"),
+        pytest.lazy_fixture("data_1d_bad_shape"),
+    ],
+)
+def test_equations(data):
+    x, t = data
+    model = SINDy()
+    model.fit(x, t)
+
+    model.print(precision=2)
```
[the remark this answers is not in the document]
Comment (briandesilva) [reply test/test_pysindy.py]: Updated in latest commit

File: test/optimizers/test_optimizers.py [hunk 4]
```diff
@@ -31,6 +34,9 @@ def test_alternate_parameters(data_derivative_1d):
     model = STLSQ(copy_X=False)
     model.fit(x, x_dot)
 
+    model = STLSQ()
+    model.fit(x, x_dot, sample_weight=x[:, 0])
```
[the remark this answers is not in the document]
Comment (briandesilva) [reply test/optimizers/test_optimizers.py]: Updated in latest commit

File: pysindy/optimizers/elastic_net.py [hunk 5]
```diff
@@ -91,3 +93,8 @@ def _reduce(self, x, y):
 
         self.coef_ = elastic_net_model.coef_
         self.iters = elastic_net_model.n_iter_
+
+    @property
+    def complexity(self):
+        check_is_fitted(self)
+        return count_nonzero(self.coef_)
```
Comment (Ohjeah) [inline pysindy/optimizers/elastic_net.py]: What about the intercept (if `fit_intercept=True`)?

Comment (briandesilva) [reply pysindy/optimizers/elastic_net.py]: I've added an extra term for the intercept, but in doing so I uncovered another problem. I'll create an issue and elaborate there.

File: pysindy/optimizers/stlsq.py [hunk 6]
```diff
@@ -147,3 +147,11 @@ def _reduce(self, x, y):
                 )
         self.coef_ = coef
         self.ind_ = ind
+
+    @property
+    def complexity(self):
+        raise NotImplementedError
+        # TODO: add check for optimizer being fitted
```
[the remark this answers is not in the document]
Comment (briandesilva) [reply pysindy/optimizers/stlsq.py]: Fixed

**Pull request merged**